beets/dbcore/db.py

```diff
@@ -1059,43 +1076,155 @@ def _make_attribute_table(self, flex_table):
-        # Doing the per-item filtering in python is faster than issuing
-        # one query per item to sqlite.
-        flex_sql = ("""
-            SELECT * FROM {} WHERE entity_id IN
-                (SELECT id FROM {} WHERE {});
-            """.format(
-            model_cls._flex_table,
-            model_cls._table,
-            where or '1',
-        )
-        )
+        fields = defaultdict(set)
+        for field, fast in self._get_fields(query):
+            fields[fast].add(field)
+
+        flex_fields, model_fields = fields[False], fields[True]
+        relation_fields = model_fields - set(model_cls._fields)
+
+        table = model_cls._table
+        _from = table
+        # select all fields from the queried entity
+        select_fields = [f"{table}.*"]
+        if flex_fields:
+            # prefetch IDs of entities that match the filter which includes
+            # flexible attributes
+            ids = self._get_matching_ids(model_cls, where or 1, subvals)
+            where = f"{table}.id IN ({', '.join(map(str, ids))})"
+            subvals = []
+        elif relation_fields:
+            # otherwise, if we are filtering by a related field, join
+            # the related entity table, and return the required fields,
+            # such as `items.path`
+            _from = self._relation_join(model_cls)
+            select_fields += relation_fields
+
+        sql = f"""
+        SELECT
+            {", ".join(select_fields)},
+            group_to_json(key, value) AS "flex_attrs [json_str]"
+        FROM {_from}
+        LEFT JOIN (
+            SELECT entity_id, key, value
+            FROM {model_cls._flex_table}
+        ) ON {table}.id = entity_id
+        WHERE {where or 1}
+        GROUP BY {table}.id
+        """
+        if order_by:
+            # a field name in `order_by`, say `album`, would be ambiguous
+            # in the `sql` query since it may exist on both (joined) tables
+            # in `_from`, causing sqlite3.OperationalError.
+            # Since we know that `sql` selects unique fields, we wrap it in
+            # a subquery.
+            # This is also good for performance, since ordering is applied
+            # to the final (filtered) list of entities, staying away from the
+            # joins and the GROUP BY clause.
+            sql = f"SELECT * FROM ({sql}) ORDER BY {order_by}"
+        if limit:
+            # use the limit at the end ensuring that we limit sorted entities
+            sql += f"\nLIMIT {limit}"
+
+        self.print_query(sql, subvals)
 
         with self.transaction() as tx:
             rows = tx.query(sql, subvals)
-            flex_rows = tx.query(flex_sql, subvals)
 
         return Results(
-            model_cls, rows, self, flex_rows,
-            None if where else query,  # Slow query component.
+            model_cls, rows, self,
             sort if sort.is_slow() else None,  # Slow sort component.
```

Comment:
This should probably also go actually.

Reply:
Attempted making sorting 'quick' but found that it is not at all straightforward and that it's already well optimized.